### PR TITLE
SYNCT-75: Fixed issue with synchronization loop, during object update operation.

### PR DESCRIPTION
**Pull request by pkornowski**

### CODE REVIEW

File: api/src/main/java/org/openmrs/module/sync2/client/rest/resource/Location.java [hunk 1]
```diff
@@ -268,4 +269,44 @@ public BaseOpenmrsObject getOpenMrsObject() {
...
+                Objects.equals(address2, location.address2) &&
+                Objects.equals(address3, location.address3) &&
+                Objects.equals(address4, location.address4) &&
+                Objects.equals(address5, location.address5) &&
+                Objects.equals(address6, location.address6) &&
+                Objects.equals(tags, location.tags) &&
+                Objects.equals(parentLocation, location.parentLocation) &&
+                Objects.equals(childLocations, location.childLocations) &&
+                Objects.equals(retired, location.retired) &&
+                Objects.equals(links, location.links) &&
```
Comment (dserkowski) [inline api/src/main/java/org/openmrs/module/sync2/client/rest/resource/Location.java]: Should we compare links? Look at comments above.

File: api/src/main/java/org/openmrs/module/sync2/client/rest/resource/Location.java [hunk 2]
```diff
@@ -268,4 +269,44 @@ public BaseOpenmrsObject getOpenMrsObject() {
...
+                Objects.equals(resourceVersion, location.resourceVersion);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(uuid, display, name, description, cityVillage, stateProvince, country, postalCode, latitude, longitude, countryDistrict, address1, address2, address3, address4, address5, address6, tags, parentLocation, childLocations, retired, links, resourceVersion);
```
Comment (dserkowski) [inline api/src/main/java/org/openmrs/module/sync2/client/rest/resource/Location.java]: Should we compare links? Look at comments above.

File: api/src/main/java/org/openmrs/module/sync2/client/rest/resource/LocationTag.java [hunk 3]
```diff
@@ -76,4 +77,28 @@ public BaseOpenmrsObject getOpenMrsObject() {
...
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+
+        LocationTag that = (LocationTag) o;
+        return Objects.equals(uuid, that.uuid) &&
+                Objects.equals(name, that.name) &&
+                Objects.equals(description, that.description) &&
+                Objects.equals(retired, that.retired) &&
+                Objects.equals(links, that.links) &&
```
Comment (dserkowski) [inline api/src/main/java/org/openmrs/module/sync2/client/rest/resource/LocationTag.java]: Should we compare links? Look at comments above.

File: api/src/main/java/org/openmrs/module/sync2/client/rest/resource/LocationTag.java [hunk 4]
```diff
@@ -76,4 +77,28 @@ public BaseOpenmrsObject getOpenMrsObject() {
...
+                Objects.equals(resourceVersion, that.resourceVersion);
+    }
+
+    @Override
+    public int hashCode() {
+
+        return Objects.hash(uuid, name, description, retired, links, resourceVersion);
```
Comment (dserkowski) [inline api/src/main/java/org/openmrs/module/sync2/client/rest/resource/LocationTag.java]: Should we compare links? Look at comments above.

File: api/src/main/java/org/openmrs/module/sync2/client/rest/resource/LocationTag.java [hunk 5]
```diff
@@ -76,4 +77,28 @@ public BaseOpenmrsObject getOpenMrsObject() {
...
+                Objects.equals(name, that.name) &&
+                Objects.equals(description, that.description) &&
+                Objects.equals(retired, that.retired) &&
+                Objects.equals(links, that.links) &&
+                Objects.equals(resourceVersion, that.resourceVersion);
+    }
+
+    @Override
+    public int hashCode() {
+
```
Comment (dserkowski) [inline api/src/main/java/org/openmrs/module/sync2/client/rest/resource/LocationTag.java]: This empty line is unnecessary

File: api/src/main/java/org/openmrs/module/sync2/client/rest/resource/Person.java [hunk 6]
```diff
@@ -170,4 +171,57 @@ public Date getBirthtime() {
...
+                Objects.equals(addresses, person.addresses) &&
+                Objects.equals(birthdateEstimated, person.birthdateEstimated) &&
+                Objects.equals(age, person.age) &&
+                Objects.equals(gender, person.gender) &&
+                Objects.equals(dead, person.dead) &&
+                Objects.equals(causeOfDeath, person.causeOfDeath) &&
+                Objects.equals(deathDate, person.deathDate) &&
+                Objects.equals(voided, person.voided) &&
+                Objects.equals(deathdateEstimated, person.deathdateEstimated);
+    }
```
Comment (dserkowski) [inline api/src/main/java/org/openmrs/module/sync2/client/rest/resource/Person.java]: I think that you should create a ticket and comment why you don't compare birthdate and birthtime

File: api/src/main/java/org/openmrs/module/sync2/client/rest/resource/PersonName.java [hunk 7]
```diff
@@ -107,4 +108,31 @@ public BaseOpenmrsObject getOpenMrsObject() {
...
+                Objects.equals(voided, that.voided) &&
+                Objects.equals(links, that.links) &&
+                Objects.equals(resourceVersion, that.resourceVersion);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(uuid, display, givenName, middleName, familyName, familyName2, voided, links, resourceVersion);
+    }
 }
```
Comment (dserkowski) [inline api/src/main/java/org/openmrs/module/sync2/client/rest/resource/PersonName.java]: In other classes, you created toString method but here not. Why?

File: api/src/main/java/org/openmrs/module/sync2/client/rest/resource/PersonName.java [hunk 8]
```diff
@@ -107,4 +108,31 @@ public BaseOpenmrsObject getOpenMrsObject() {
...
+
+        PersonName that = (PersonName) o;
+        return Objects.equals(uuid, that.uuid) &&
+                Objects.equals(display, that.display) &&
+                Objects.equals(givenName, that.givenName) &&
+                Objects.equals(middleName, that.middleName) &&
+                Objects.equals(familyName, that.familyName) &&
+                Objects.equals(familyName2, that.familyName2) &&
+                Objects.equals(voided, that.voided) &&
+                Objects.equals(links, that.links) &&
```
Comment (dserkowski) [inline api/src/main/java/org/openmrs/module/sync2/client/rest/resource/PersonName.java]: Should we compare links? Look at comments above.

File: api/src/main/java/org/openmrs/module/sync2/client/rest/resource/PersonName.java [hunk 9]
```diff
@@ -107,4 +108,31 @@ public BaseOpenmrsObject getOpenMrsObject() {
...
+                Objects.equals(resourceVersion, that.resourceVersion);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(uuid, display, givenName, middleName, familyName, familyName2, voided, links, resourceVersion);
```
Comment (dserkowski) [inline api/src/main/java/org/openmrs/module/sync2/client/rest/resource/PersonName.java]: Should we compare links? Look at comments above.

File: api/src/main/java/org/openmrs/module/sync2/client/rest/resource/Address.java [hunk 10]
```diff
@@ -264,4 +265,37 @@ public Address create() {
...
+                Objects.equals(address6, address.address6) &&
+                Objects.equals(startDate, address.startDate) &&
+                Objects.equals(endDate, address.endDate) &&
+                Objects.equals(latitude, address.latitude) &&
+                Objects.equals(longitude, address.longitude);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(uuid, address1, address2, cityVillage, stateProvince, country, postalCode, countyDistrict, address3, address4, address5, address6, startDate, endDate, latitude, longitude);
```
[the remark this answers is not in the document]
Comment (pkornowski) [reply api/src/main/java/org/openmrs/module/sync2/client/rest/resource/Address.java]: Corrected.

File: api/src/main/java/org/openmrs/module/sync2/client/rest/resource/Identifier.java [hunk 11]
```diff
@@ -110,4 +111,44 @@ public BaseOpenmrsObject getOpenMrsObject() {
...
+        }
+
+        Identifier that = (Identifier) o;
+        return Objects.equals(uuid, that.uuid) &&
+                Objects.equals(identifier, that.identifier) &&
+                Objects.equals(identifierType, that.identifierType) &&
+                Objects.equals(location, that.location) &&
+                Objects.equals(preferred, that.preferred) &&
+                Objects.equals(voided, that.voided) &&
+                Objects.equals(links, that.links);
```
[the remark this answers is not in the document]
Comment (pkornowski) [reply api/src/main/java/org/openmrs/module/sync2/client/rest/resource/Identifier.java]: You're right, corrected.

File: api/src/main/java/org/openmrs/module/sync2/client/rest/resource/Person.java (hunk 6 above)
Comment (pkornowski) [reply api/src/main/java/org/openmrs/module/sync2/client/rest/resource/Person.java]: Yes, I'll do that.

**Comment by pkornowski** [comment]
@dserkowski 
I'll create the ticket for the expose annotation, it is not in the scope of this ticket.

File: api/src/main/java/org/openmrs/module/sync2/client/rest/resource/LocationTag.java [hunk 12]
```diff
@@ -76,4 +77,26 @@ public BaseOpenmrsObject getOpenMrsObject() {
...
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+
+        LocationTag that = (LocationTag) o;
+        return Objects.equals(uuid, that.uuid) &&
+                Objects.equals(name, that.name) &&
+                Objects.equals(description, that.description) &&
+                Objects.equals(retired, that.retired) &&
+                Objects.equals(links, that.links);
```
Comment (dserkowski) [inline api/src/main/java/org/openmrs/module/sync2/client/rest/resource/LocationTag.java]: Should we compare links?

File: api/src/main/java/org/openmrs/module/sync2/client/rest/resource/Location.java [hunk 13]
```diff
@@ -268,4 +269,45 @@ public BaseOpenmrsObject getOpenMrsObject() {
...
+                Objects.equals(address2, location.address2) &&
+                Objects.equals(address3, location.address3) &&
+                Objects.equals(address4, location.address4) &&
+                Objects.equals(address5, location.address5) &&
+                Objects.equals(address6, location.address6) &&
+                Objects.equals(tags, location.tags) &&
+                Objects.equals(parentLocation, location.parentLocation) &&
+                Objects.equals(childLocations, location.childLocations) &&
+                Objects.equals(retired, location.retired) &&
+                Objects.equals(links, location.links);
```
Comment (dserkowski) [inline api/src/main/java/org/openmrs/module/sync2/client/rest/resource/Location.java]: Should we compare links?

**Pull request merged**